NeoML/src/TraditionalML/SMOptimizer.cpp

```diff
@@ -495,30 +539,166 @@ void CSMOptimizer::optimizePair( int i, int j, CArray<double>& alpha, CArray<dou
 			}
 		}
 	}
-	// Modify the gradient
+	
+	// Modify the g
 	double deltaAlpha_i = alpha[i] - oldAlpha_i;
 	double deltaAlpha_j = alpha[j] - oldAlpha_j;
-	for(int k = 0; k < data->GetVectorCount(); k++) {
-		gradient[k] += Q_i[k] * deltaAlpha_i + Q_j[k] * deltaAlpha_j;
+	for(int k = 0; k < activeSize; k++) {
+		g[k] += Q_i[k] * deltaAlpha_i + Q_j[k] * deltaAlpha_j;
+	}
+}
+
+void CSMOptimizer::updateAlphaStatusAndGradient0( int i )
+{
+	double C_i = C[i];
+	bool wasUB = alphaStatus[i] == AS_UpperBound;
+
+	if( alpha[i] >= C_i ) {
+		alphaStatus[i] = AS_UpperBound;
+	} else if( alpha[i] <= 0 ) {
+		alphaStatus[i] = AS_LowerBound;
+	} else {
+		alphaStatus[i] = AS_Free;
+	}
+
+	bool isUB = alphaStatus[i] == AS_UpperBound;
+	if( wasUB != isUB ) {
+		auto Q_i = Q->GetColumn( i, l );
+		if( wasUB ) {
+			for( int k = 0; k < l; ++k ) {
+				g0[k] -= C_i * Q_i[k];
+			}
+		} else {
+			for( int k = 0; k < l; ++k ) {
+				g0[k] += C_i * Q_i[k];
+			}
+		}
 	}
 }
 
+// reconstruct inactive elements of G from G_bar and free variables
+void CSMOptimizer::reconstructGradient()
+{
+	if( activeSize == l ) {
+		return;
+	}
+
+	int freeCount = 0;
+	for( int j = activeSize; j < l; ++j ) {
+		g[j] = g0[j] - 1;
+	}
+
+	for( int j = 0; j < activeSize; ++j ) {
+		if( alphaStatus[j] == AS_Free ) {
+			++freeCount;
+		}
+	}
+
+	if( log != nullptr && 2 * freeCount < activeSize ) {
+		*log << "\nWarning: using Shrinking=false may be faster\n";
+	}
+
+	if( freeCount * l > 2 * activeSize * ( l - activeSize ) )
+	{
+		for( int i = activeSize; i < l; ++i ) {
+			auto Q_i = Q->GetColumn( i, activeSize );
+			for( int j = 0; j < activeSize; ++j ) {
+				if( alphaStatus[j] == AS_Free ) {
+					g[i] += alpha[j] * Q_i[j];
+				}
+			}
+		}
+	} else {
+		for( int i = 0; i < activeSize; ++i ) {
+			if( alphaStatus[i] == AS_Free ) {
+				auto Q_i = Q->GetColumn( i, l );
+				double alpha_i = alpha[i];
```

Comment:
Do we really need this variable?

Reply:
what variable? alpha_i? it's all about micro-optimizations that merely can be noticed separately but collectively bring a valuable impact